rfcs/20190509-keras-saved-model.md

```diff
@@ -0,0 +1,228 @@
+```
+
+And later, loading the SavedModel:
+
+```
+model = tf.keras.load_model('/tmp/keras_model')
+model.signatures  #  {'classify' → tf.function, 'predict' → tf.function} 
+```
+
+### Serialization coverage
+
+The goal is simple - save a Keras layer or model object to SavedModel, and load it back in the same state. Same state is defined as all of the below are equivalent- 
+- variables (list of all variables, trainable variables, and non-trainable variables)
+- forward pass (call function)
+- losses
+- metrics
+- child layers (and their states)
+- python attributes
+  - name
+  - trainable
+  - config
+  - dtype
+  - batch_input_shape 
+  - input_spec
+- (if compiled model) compiled arguments
+  - optimizer (hyperparameters, iterations)
+  - loss functions
+  - metrics
+
+All of the above attributes are serialized in the SavedModel as either metadata, checkpointable objects, or  functions.
+
+Checkpointable objects and functions are as defined in [RFC: SavedModel Save/Load in 2.x](https://github.com/tensorflow/community/blob/master/rfcs/20181116-saved-model.md) (see *Concrete function graph* and *Checkpointable* in the Terminology section).
+
+Metadata will be a newly added property of checkpointable objects in the SavedModel. Checkpointable objects may save arbitrary strings to this field. Python attributes such as layer names are saved to the metadata.
+
+See [Detailed Design](#detailed-design) for more info on how each attribute is saved. 
+
+### Model deserialization
+Reconstructing a Keras model is a matter of remapping the saved attributes (listed above) to the original attributes. A few more details are in the [Detailed design](#detailed-design).
+
+## Detailed Design
+
+### Serialization details
+Keras models contain many Python-defined components. For example, the `trainable_variable` property lists the model's trainable variables by recursively retrieving the trainable variables from each of the child layers. Another example is `model.call`, a python function that calls child layers and adds ops to the backend graph.
+
+Only Tensorflow checkpointable objects and functions can be serialized to SavedModel. Serializing a Keras model as-is results in a checkpointable object that does not resemble a Keras model at all. Thus, extra checkpointable objects and functions must be created during serialization. 
+
+The following checkpointable objects and functions are attached to the saved Keras model:
+
+- `variables`: List of all variables in this layer and submodules.
+- `trainable_variables`: List of all trainable variables in this layer and submodules.
+- `non_trainable_variables`: List of all non-trainable variables in this layer and submodules.
+- `regularization_losses`: List of unconditional loss functions in this layer 
+  and submodules. Each function takes no arguments, and returns a scalar 
+  tensor. 
+- `submodules`: Flat list of all sublayers (does not include metrics, even though Metric subclasses Layer).
```

Comment:
We don't track modules, only layers, so this should be `layers`

Reply:
Done